src/main/java/org/spongepowered/api/item/recipe/crafting/CraftingRecipe.java

```diff
@@ -0,0 +1,93 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.crafting;
+
+import org.spongepowered.api.item.ItemTypes;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.item.recipe.Recipe;
+import org.spongepowered.api.world.World;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * <p>A CraftingRecipe represents some craftable recipe in the game.</p>
+ *
+ * <p>It is essentially a Predicate that checks for if a recipe is valid as well
+ * as a function from a crafting matrix to a list of {@link ItemStack}
+ * (the crafting result), therefore making it an immutable interface.</p>
+ *
+ * <p>The passed in ItemGrid is usually a crafting inventory, e.g.
+ * a 2x2 or 3x3 crafting matrix.</p>
+ *
+ * <p>The requirements of a CraftingRecipe can be general, they just have to
+ * eventually return a boolean given an itemgrid.</p>
+ *
+ * <p>You can implement it manually to suit your creative needs,
+ * or you can simply use the {@link ShapelessCraftingRecipe.Builder} or {@link ShapedCraftingRecipe.Builder}.</p>
+ */
+public interface CraftingRecipe extends Recipe {
+    /**
+     * Checks if the given {@link GridInventory} fits the required constraints
+     * to craft this CraftingRecipe.
+     *
+     * @param grid The ItemGrid to check for validity
+     * @return True if the given input matches this recipe's requirements
+     */
+    boolean isValid(GridInventory grid, World world);
+
+    /**
+     * This method is preferred over the {@link #getExemplaryResult()} method,
+     * as it customizes the result further depending on the specified
+     * input {@param grid}.
+     * It is advised to use the output of {@link #getExemplaryResult()},
+     * modify it accordingly, and {@code return} it.
+     *
+     * @param grid The crafting input, typically 3x3 or 2x2
+     * @return An {@link ItemStack} or {@link Optional#empty()} if the given
+     *         {@link GridInventory} does not match this recipe's requirements.
+     */
+    Optional<ItemStack> getResult(GridInventory grid);
```

Comment:
Here you get the results but how do you actually consume the items? Is it required to reduce the number by one? How about Blood Magic's magicians orbs that act as a mere catalyst and thus is not consumed?  Or recipes where a tool is used and the tool only takes some damage?